### PR TITLE
New convenience cython methods for SpanGroup

### PR DESCRIPTION
**Pull request by nrodnova**
Added new convenience cython methods for `SpanGroup` to avoid unnecessary copying of the underlying data. Also, updated documentation accordingly. Added tests for `SoanGroup` class.

## Description

Added functions:
We discussed a bit `__setitem__`, but came to no conclusion. I threw it, along with `__delitem__` into the pile. I can delete both if you don't like it. 
* `__delitem__(self, int i)`
* `__setitem__(self, int i, Span span)`

The following are efficient c++ versions of functions I implemented outside of `spaCy` and found myself using a lot:
* `sort(self, inplace = False, longest_first = False)` - self-explanatory. Provides an option to prefer longest spans (if start is the same), consistent with `util.filter_spans`
* `clone(self)` - self-explanatory
* `def filter_spans(self, inplace = False)` - a c++ version of `util.filter_spans`
* `get_overlaps(self, Span span, exclude_self = True, exclude_partial = False)` - returns a `SpanGroup` consisting of spans overlapping with the given one. Useful in cases when you want to get spans in a given sentence, for example.
* `concat(self, other : Union[SpanGroup, Iterable["Span"]], sort_spans = False, filter_spans = False, inplace = False) ` - appending 2 groups with an option to sort and filter the result

A non-member function allowing to concatenate multiple `SpanGroups`:
*  `concat_span_groups(span_groups : Iterable[SpanGroup], name = None, attrs = None, sort_spans = False, filter_spans = False)` - useful to flatten `doc.spans` to be used in `SpanGroup.get_overlaps`
* 
And some syntactic sugar:
* ` __iadd__(self, other : Union[SpanGroup, Iterable["Span"]])`  - an equivalent of `self.concat` with `inplace = True`
* `__add__(self, other : Union[SpanGroup, Iterable["Span"]])` - an equivalent of `self.concat` with `inplace = False`

### Types of change

enhancements, new tests, documentation

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by adrianeboyd** [comment]
Thanks for this PR! It's nice to see how the span groups are getting used in practice. We want to discuss the details a bit, but we'll try to get back to you by next week. We also still have the `shared_ptr` idea on our list of things to take a look at, sorry for the delay on that!

**Comment by nrodnova** [comment]
@adrianeboyd Thank you and no rush. If you decide to go with it, I'll be happy to make any changes you think are necessary.  I also ran wild with tests here, I didn't realize that the size of the test suite was an issue. I can definitely tighten it up. 

And I hope you embrace the `shared_ptr` idea, unless there is a better one :) I had another thought, actually - maybe having `Span` be able to hold a weak and a strong links to the doc. And when a `Span` goes into the `SpanGroup`, it loses the strong link, and whenever it is returned by getters, it gets the strong link? It sounds a bit hairy, but it could be nicely incapsulated. Probably better that creating and deleting `Spans` all the time.

File: spacy/tests/util.py [hunk 1]
```diff
@@ -94,3 +94,24 @@ def assert_packed_msg_equal(b1, b2):
     for (k1, v1), (k2, v2) in zip(sorted(msg1.items()), sorted(msg2.items())):
         assert k1 == k2
         assert v1 == v2
+
+def assert_span_list_equal(spans_1, spans_2) :
+    spans_1 = list(spans_1)
+    spans_2 = list(spans_2)
```
Comment (svlandeg) [inline spacy/tests/util.py:100]: If these are not assumed to be lists on input, shouldn't they be sorted as well before zipping and comparing element by element?

File: spacy/tests/doc/test_span_group.py [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+
+    span_group_1 = doc.spans['SPANS']
+    span_group_1.extend(spans)
+    assert len(span_group_1) == len(span_group_1_expected)
+    assert span_group_1.attrs == {'key': 'value'}
+    assert_span_list_equal(span_group_1, span_group_1_expected)
+
+def test_span_group_dealloc(span_group) :
+    with pytest.raises(AttributeError) :
+        print(span_group.doc)
```
Comment (svlandeg) [inline spacy/tests/doc/test_span_group.py:401]: Generic note: this file still requires `black` formatting (I know you know the drill by now ;-))

File: spacy/tokens/span_group.pyx [hunk 3]
```diff
@@ -4,7 +4,13 @@ import srsly
...
 from spacy.errors import Errors
 from .span cimport Span
-from libc.stdint cimport uint64_t, uint32_t, int32_t
+from ..structs cimport SpanC
+
+from libcpp.algorithm cimport sort as sort_vector
+from libcpp.set cimport set
+from libcpp.utility cimport pair
+
+from typing import Iterable, Tuple, Union, Optional, TYPE_CHECKING
```
Comment (svlandeg) [inline spacy/tokens/span_group.pyx:13]: You couldn't have known this, but we typically put the imports from `typing` at the very first line

File: spacy/tokens/span_group.pyx [hunk 4]
```diff
@@ -101,17 +111,15 @@ cdef class SpanGroup:
             raise ValueError("Cannot add span to group: refers to different Doc.")
         self.push_back(span.c)
 
-    def extend(self, spans):
-        """Add multiple spans to the group. All spans must refer to the same
-        Doc object as the span group.
+    def extend(self, spans_or_span_group : Union[SpanGroup, Iterable["Span"]]):
+        """Add multiple spans or contents of another SoanGroup to the group.
```
Comment (svlandeg) [inline spacy/tokens/span_group.pyx:115]:
```suggestion
        """Add multiple spans or contents of another SpanGroup to the group.
```

File: spacy/tokens/span_group.pyx [hunk 5]
```diff
@@ -128,6 +136,65 @@ cdef class SpanGroup:
...
+
+        i (int): The item index.
+
+        DOCS: https://spacy.io/api/spangroup#delitem
+        """
+        cdef int size = self.c.size()
+        if i < -size or i >= size:
+            raise IndexError(f"list index {i} out of range")
+        if i < 0:
+            i += size
```
Comment (svlandeg) [inline spacy/tokens/span_group.pyx:150]: It looks like we're copying the check & adjustement of index `i` across various functions - perhaps we could use a simple private helper for this?

File: website/docs/api/spangroup.md [hunk 6]
```diff
@@ -104,7 +104,7 @@ Get the number of spans in the group.
 
 ## SpanGroup.\_\_getitem\_\_ {#getitem tag="method"}
 
-Get a span from the group.
+Get a span from the group. Note, that a copy of the span is returned, so if any changes are made to this span, they are not going to be reflected in the corresponsing member of the group. Similarly to `doc.ents`, the group will need to be re-assigned for changes to be applied.
```
Comment (svlandeg) [inline website/docs/api/spangroup.md:107]:
```suggestion
Get a span from the group. Note, that a copy of the span is returned, so if any changes are made to this span, they are not going to be reflected in the corresponding member of the group. The group will need to be re-assigned for changes to be applied.
```

File: website/docs/api/spangroup.md [hunk 7]
```diff
@@ -113,13 +113,56 @@ Get a span from the group.
...
+> assert doc.spans["errors"][1] != 'LABEL' # The span within the group was not updated
 > ```
 
 | Name        | Description                           |
 | ----------- | ------------------------------------- |
 | `i`         | The item index. ~~int~~               |
 | **RETURNS** | The span at the given index. ~~Span~~ |
 
+
+## SpanGroup.\_\_setitem\_\_ {#setitem tag="method"}
```
Comment (svlandeg) [inline website/docs/api/spangroup.md:126]: The new methods could get a `new="3.2.1"` marker for the new methods, and the docs file should be formatted with `prettier` (again, I know you know from previous PRs, but just making sure we don't forget it here).

**Review by svlandeg — COMMENTED** [review]
Some quick first comments

File: spacy/tests/util.py (hunk 1 above)
Comment (nrodnova) [reply spacy/tests/util.py]: The order in lists is important when comparing. I overdid the type conversion there, for sure. I actually changed the function from asserting to simply comparing, to eliminate the need of a separate negative assertion helper.

File: spacy/tests/doc/test_span_group.py (hunk 2 above)
Comment (nrodnova) [reply spacy/tests/doc/test_span_group.py]: I set up automatic formatting according to the instructions you gave me. All good now :)

File: spacy/tokens/span_group.pyx (hunk 3 above)
Comment (nrodnova) [reply spacy/tokens/span_group.pyx]: Done :)

File: spacy/tokens/span_group.pyx (hunk 5 above)
Comment (nrodnova) [reply spacy/tokens/span_group.pyx]: Done.

File: website/docs/api/spangroup.md (hunk 7 above)
Comment (nrodnova) [reply website/docs/api/spangroup.md]: I added the `new` tag. Still haven't figured out how to make Prettier use spacy's format settings in PyCharm. I followed the plug-in instructions, but no luck - it doesn't respect settings in `.prettierrc` So no prettier for now, sorry. If someone could help with it, would be great.

Comment (svlandeg) [reply website/docs/api/spangroup.md]: Oh, that's weird, I run `prettier` from within Pycharm. I don't think I had to edit anything specific in `.prettierrc`. But getting `prettier` to run can sometimes be a bit of a hassle. Let's not worry about it for now. If I can push to your branch directly I can look into it later.

File: spacy/tests/doc/test_span_group.py [hunk 8]
```diff
@@ -0,0 +1,576 @@
...
+    )
+    matcher.add(
+        "1",
+        [
+            [
+                {},
+            ]
+        ],
+    )
+
```
Comment (svlandeg) [inline spacy/tests/doc/test_span_group.py:32]:
```suggestion
# fmt: off
    matcher.add("4", [[{}, {}, {}, {}]])
    matcher.add("2", [[{}, {}, ]])
    matcher.add("1", [[{}, ]])
# fmt: on
```
Sorry, I should have also explained that sometimes we "cheat" when the linebreaks intervene with readibility ;-)

File: spacy/tests/doc/test_span_group.py [hunk 9]
```diff
@@ -0,0 +1,576 @@
...
+        ],
+    )
+    matcher.add(
+        "1",
+        [
+            [
+                {},
+            ]
+        ],
+    )
```
Comment (svlandeg) [inline spacy/tests/doc/test_span_group.py:68]:
```suggestion
# fmt: off
    matcher.add("4", [[{}, {}, {}, {}]])
    matcher.add("2", [[{}, {}, ]])
    matcher.add("1", [[{}, ]])
# fmt: on
```

File: spacy/tests/doc/test_span_group.py [hunk 10]
```diff
@@ -0,0 +1,576 @@
...
```
Comment (svlandeg) [inline spacy/tests/doc/test_span_group.py:105]:
```suggestion
# fmt: off
    matcher.add("4", [[{}, {}, {}, {}]])
    matcher.add("2", [[{}, {}, ]])
    matcher.add("1", [[{}, ]])
# fmt: on
```

**Review by svlandeg — COMMENTED** [review]
Thanks for the updates! I know Matt wanted to review this in a bit more detail as well.

Also, I should have explained that sometimes we "cheat" when the linebreaks intervene with readibility. Apologies for the nitpicking ;-)

File: website/docs/api/spangroup.md (hunk 7 above)
Comment (nrodnova) [reply website/docs/api/spangroup.md]: I just noticed that I get an error message in PyCharm's log:
```
7:38 AM	Prettier: please specify Node.js interpreter
Configure Node.js path...
```
That's probably the reason. It does format now, but it re-formats things that I haven't touched, and I can see that the lines are wider than there were originally.

I will have to take a look at some point.

Comment (svlandeg) [reply website/docs/api/spangroup.md]: Ah, hm. I do know that there can be incompatibilities between the `node` and `npm` versions. I typically pin them at  `node` v11.15.0 & `npm` 6.14.4. Then in Pycharm you can double check your Node.js interpreter and Package manager, and Prettier package settings.

Comment (svlandeg) [reply website/docs/api/spangroup.md]: (but again, if it's too much hassle, we'll do it for you ;-))

**Comment by adrianeboyd** [comment]
Just taking another look at this since the span comparison and `shared_ptr` PRs have been merged.

We think it makes sense to:

* support the magic methods `__setitem__`, `__delitem__`, `__add__`, and `__iadd__` in `SpanGroup`
* add private helper methods like `SpanGroup._concat` as needed
* add a very small number of general-purpose cython utils for sorting and filtering `vector[SpanC]`

The filtering methods have been a source of bugs and people often want different customizations anyway, so I think we should try to limit the number of methods/options that we maintain in the core library.

There are two places where the same basic `util.filter_spans` filtering is implemented in python (which is probably one too many and I'd have to look again at the details to remember why we kept both):

* `util.filter_spans`
* as part of `EntityRuler.set_annotations`

And one place with filtering in cython (which is customized for the tokenizer):

* `Tokenizer._filter_special_spans` plus two comparison methods (one of which is poorly named)
   https://github.com/explosion/spaCy/blob/2ff53834bb09eea2af3b7715a2516bcf7913a370/spacy/tokenizer.pyx#L270-L284

I think we could move the tokenizer methods into a more general misc/utils location and use them for both the tokenizer and span groups.

**Comment by svlandeg** [comment]
Just as a note, we still want to work on this further, but we're dealing with some other priorities right now. Putting this in draft until we can get to it :-)

**Comment by nrodnova** [comment]
Let me know if you want me to take care of these, I missed the comment from January 21 somehow...

**Comment by svlandeg** [comment]
I think Adriane will spend some time implementing the changes she outline - it's mostly about repo organization really!

**Comment by svlandeg** [comment]
Closing in favour of #10476, where the commits have been cherry-picked. We can continue the discussion over there :-)

**Pull request closed without merging**